File: languages/typescript/protobuf/bench.ts
// Protobuf TypeScript benchmark target.
//
// Encodes + decodes the canonical FullScaleExample message (schema/state.json)
// through protobufjs, loading schema/message.proto at runtime. Same message,
// same state, same timed region as the SofaBuffers target. Prints one uniform
// BENCH line (see docs/BENCH.md).
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import protobuf from "protobufjs";

// state.json stores u64/i64 as bare JSON number literals; the large ones exceed
// 2^53 and would lose precision under plain JSON.parse. Quote any 16+ digit
// integer literal so protobufjs receives an exact (string) value for its 64-bit
// fields. Float literals are untouched (their digit runs are preceded by '.').
function parseStatePreserveBigInts(text: string): Record<string, any> {
  const quoted = text.replace(/(?<![\d."])(-?\d{16,})(?![\d.])/g, '"$1"');
  return JSON.parse(quoted);
}

function build(state: Record<string, any>): Record<string, unknown> {
  const n = state.nested;
  const a = state.arrays;
  return {
    u8: state.u8, i8: state.i8, u16: state.u16, i16: state.i16,
    u32: state.u32, i32: state.i32, u64: state.u64, i64: state.i64,
    nested: {
      f32: n.f32,
      f64: n.f64,
      str: n.str,
      // protobufjs camelCases proto field names by default.
      bytesField: new Uint8Array(n.bytes_field),
    },
    arrays: {
      u8: a.u8, i8: a.i8, u16: a.u16, i16: a.i16,
      u32: a.u32, i32: a.i32, u64: a.u64, i64: a.i64,
      nested: { fp32: a.nested.fp32, fp64: a.nested.fp64 },
    },
    stringArray: { strings: state.string_array },
  };
}

async function main(): Promise<number> {
  const statePath = process.env.STATE_JSON;
  const protoPath = process.env.PROTO_PATH;
  if (!statePath || !protoPath) {
    process.stderr.write("FAIL: STATE_JSON / PROTO_PATH not set\n");
    return 1;
  }
  const root = await protobuf.load(protoPath);
  const Type = root.lookupType("fullscale.FullScaleExample");

  const state = parseStatePreserveBigInts(readFileSync(statePath, "utf8"));
  const payload = build(state);
  // fromObject coerces string 64-bit literals into Long, preserving exact
  // values beyond 2^53 (Type.verify would reject the string form).
  const src = Type.fromObject(payload);

  // Warm-up round-trip + self-check (outside the timed region).
  const blob = Type.encode(src).finish();
  const serialized = blob.length;
  const sha = createHash("sha256").update(blob).digest("hex");
  const reencoded = Type.encode(Type.decode(blob)).finish();
  if (reencoded.length !== blob.length || !blob.every((b, i) => b === reencoded[i])) {
    process.stderr.write("FAIL: protobuf round-trip self-check\n");
    return 1;
  }

  const iters = parseInt(process.env.BENCH_ITERS ?? "500000", 10);

  // Warm the JIT.
  for (let i = 0; i < 10000; i++) {
    Type.decode(Type.encode(src).finish());
  }

  const t0 = process.hrtime.bigint();
  for (let i = 0; i < iters; i++) {
    Type.decode(Type.encode(src).finish());
  }
  const t1 = process.hrtime.bigint();

  const cpu = Number(t1 - t0) / 1e9;
  const mbs = cpu > 0 ? (serialized * iters) / cpu / 1e6 : 0;
  process.stdout.write(
    `BENCH lang=typescript impl=protobuf serialized_bytes=${serialized} ` +
      `iters=${iters} cpu_time_s=${cpu.toFixed(6)} ` +
      `throughput_mbs=${mbs.toFixed(2)} sha256=${sha}\n`,
  );
  return 0;
}

main().then((c) => process.exit(c));

File: languages/csharp/sofab/bench/Bench.cs
// SofaBuffers C# benchmark target.
// Encodes + decodes the canonical Example message, hand-filled from
// schema/STATE.md, through the sofabgen-generated Sofabuffers types. Same
// message, same state, same timed region as the Protobuf target.
//
// ONE source, TWO impls, selected by the BENCH_IMPL env var (#108):
//
//   sofab         Encode() — one buffer the size of the whole message
//   sofab-stream  EncodeTo(OStream) over a buffer SMALLER than the message with
//                 a FlushSink draining it as it fills, so the encode's memory
//                 need is the buffer and not the message
//
// The two paths are separate loops chosen before timing starts, never a branch
// inside the timed region. Only the ENCODE half differs; the decode stays the
// plain one-shot Decode in both, so a row reflects one axis.
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Security.Cryptography;
using sofab;          // OStream / FlushSink (corelib-cs)
using Sofabuffers;

static class Program {
    static Example Build() {
        var m = new Example {
            u8 = 200,
            i8 = -100,
            u16 = 50000,
            i16 = -20000,
            u32 = 3000000000,
            i32 = -1000000000,
            u64 = 10000000000000UL,
            i64 = -5000000000000L,
            nested = new ExampleNested {
                f32 = 3.14f,
                f64 = 3.14159265,
                str = "Hello, World!",
                bytes_field = new byte[] { 0xDE, 0xAD, 0xBE, 0xEF },
            },
            arrays = new ExampleArrays(),
            string_array = new List<string>(),
        };
        var a = m.arrays;
        a.u8 = new byte[] { 0, 64, 128, 191, 255 };
        a.i8 = new sbyte[] { -128, -64, 0, 63, 127 };
        a.u16 = new ushort[] { 0, 16384, 32768, 49151, 65535 };
        a.i16 = new short[] { -32768, -16384, 0, 16383, 32767 };
        a.u32 = new uint[] { 0, 1073741824, 2147483648, 3221225471, 4294967295 };
        a.i32 = new int[] { -2147483648, -1073741824, 0, 1073741823, 2147483647 };
        a.u64 = new ulong[] {
            0, 4611686018427387904, 9223372036854775808,
            13835058055282163711, 18446744073709551615 };
        a.i64 = new long[] {
            -9223372036854775807, -4611686018427387904, 0,
            4611686018427387903, 9223372036854775807 };
        a.nested = new ExampleArraysNested();
        a.nested.fp32 = new float[] { 1f, 2f, 3f, -float.MaxValue, float.MaxValue };
        a.nested.fp64 = new double[] { 1d, 2d, 3d, -double.MaxValue, double.MaxValue };
        m.string_array.AddRange(new string[] {
            "Hello, Sofab!", "", "1234567890", "äöüÄÖÜß",
            "This_is_a_very_long_test_string_with_!@#$%^&*()_+-=[]{}" });
        return m;
    }

    // The drained bytes land here so the loop keeps a live sink and the result
    // stays checkable against the reference wire. The encoder never sees this
    // array -- only the FlushSink does.
    static readonly byte[] Sink = new byte[2048];
    static int sinkN;

    /// <summary>Streaming encode of <paramref name="m"/>; returns bytes drained.</summary>
    static int StreamEncode(Example m, OStream os) {
        sinkN = 0;
        m.EncodeTo(os);   // Serialize(os) + os.Flush(), which also rewinds os
        return sinkN;
    }

    static int Main() {
        var src = Build();

        string impl = Environment.GetEnvironmentVariable("BENCH_IMPL") ?? "sofab";
        bool streaming = impl == "sofab-stream";
        int streamCap = int.Parse(Environment.GetEnvironmentVariable("STREAM_BUF_BYTES") ?? "64");
        OStream? os = streaming
            ? new OStream(new byte[streamCap], 0, (data, off, len) => {
                  Buffer.BlockCopy(data, off, Sink, sinkN, len);
                  sinkN += len;
              })
            : null;

        // Warm-up round-trip + self-check (outside the timed region).
        byte[] blob = src.Encode();
        int serialized = blob.Length;
        string sha = Convert.ToHexString(SHA256.HashData(blob)).ToLowerInvariant();
        // The check runs through whichever encode path this impl measures, so a
        // dropped or duplicated drained chunk fails here instead of being reported.
        bool ok = streaming
            ? StreamEncode(Example.Decode(blob), os!) == serialized
              && ((ReadOnlySpan<byte>)Sink).Slice(0, serialized).SequenceEqual(blob)
            : ((ReadOnlySpan<byte>)Example.Decode(blob).Encode()).SequenceEqual(blob);
        if (!ok) {
            Console.Error.WriteLine("FAIL: sofab round-trip self-check");
            Environment.Exit(1);
        }

        long iters = long.Parse(Environment.GetEnvironmentVariable("BENCH_ITERS") ?? "2000000");

        // JIT warm-up (same chained shape as the timed loop).
        for (int i = 0; i < 5000; i++) {
            if (streaming) StreamEncode(Example.Decode(blob), os!);
            else Example.Decode(blob).Encode();
        }

        // Chained round trip: decode the reference wire, then re-encode the freshly
        // decoded message (issue #86) — the proxy/transcode shape, which denies
        // protobuf its once-per-instance serialized-size memo so encode is measured
        // on equal terms. sink keeps the re-encode live and doubles as a loop-path
        // check (every re-encode is `serialized` bytes).
        //
        // Two loops, picked before the stopwatch starts: the impl must not cost a
        // branch per iteration that the other impl does not also pay.
        long sink = 0;
        Stopwatch sw;
        if (streaming) {
            sw = Stopwatch.StartNew();
            for (long i = 0; i < iters; i++) {
                sink += StreamEncode(Example.Decode(blob), os!);
            }
            sw.Stop();
        } else {
            sw = Stopwatch.StartNew();
            for (long i = 0; i < iters; i++) {
                sink += Example.Decode(blob).Encode().Length;
            }
            sw.Stop();
        }

        if (sink != (long)serialized * iters
            || (streaming && !((ReadOnlySpan<byte>)Sink).Slice(0, serialized).SequenceEqual(blob))) {
            Console.Error.WriteLine("FAIL: sofab loop-path self-check");
            Environment.Exit(1);
        }

        double cpu = sw.Elapsed.TotalSeconds;
        double mbs = cpu > 0 ? (double)serialized * iters / cpu / 1e6 : 0.0;
        Console.WriteLine(
            $"BENCH lang=csharp impl={impl} serialized_bytes={serialized} iters={iters} " +
            $"cpu_time_s={cpu:F6} throughput_mbs={mbs:F2} sha256={sha}");
        return 0;
    }
}
